.github/workflows/ci.yaml

```diff
@@ -1,34 +1,161 @@
 name: CI build
 
+concurrency:
+  group: ci-${{ github.run_id }}
+  cancel-in-progress: true
+
 on:
-  push:
-    branches:
-      - main
-      - v[0-9]+
-      - v[0-9]+.[0-9]+
-      - cryostat-v[0-9]+.[0-9]+
-  pull_request:
+  pull_request_target:
+    types:
+      - opened
+      - reopened
+      - synchronize
+      - labeled
+      - unlabeled
     branches:
       - main
       - v[0-9]+
       - v[0-9]+.[0-9]+
       - cryostat-v[0-9]+.[0-9]+
 
 jobs:
-  build:
+  check-before-build:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Fail if safe-to-test is not applied
+      if: github.repository_owner == 'cryostatio' && (!contains(github.event.pull_request.labels.*.name, 'safe-to-test'))
+      run: exit 1
 
+  build-core:
     runs-on: ubuntu-latest
+    outputs:
+      core_version: ${{ steps.get_core_version.outputs.core_version }}
+    steps:
+    - uses: actions/checkout@v4
+    - id: get_core_version
+      run: |
+        echo "core_version=$(mvn help:evaluate -Dexpression=project.version -q -DforceStdout)" >> $GITHUB_OUTPUT
+        echo "name=$(mvn help:evaluate -Dexpression=project.artifactId -q -DforceStdout)" >> $GITHUB_OUTPUT
+    - uses: actions/setup-java@v3
+      with:
+        java-version: '17'
+        distribution: 'temurin'
+    - uses: actions/cache@v3
+      with:
+        path: ~/.m2
+        key: ${{ runner.os }}-build-${{ hashFiles('**/pom.xml') }}
+    - run: mvn -B -U clean install
 
+  build-cryostat:
+    needs: [build-core]
+    strategy:
+      matrix:
+        arch: [amd64, arm64]
+    runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2
-    - uses: actions/setup-java@v2
+    - name: Install qemu
+      if: ${{ inputs.build-arch != 'amd64' }}
+      continue-on-error: true
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y qemu-user-static
+    - uses: actions/checkout@v4
       with:
-        java-version: '11'
+        repository: ${{ github.repository_owner }}/cryostat
+        ref: main
+        submodules: true
+        fetch-depth: 0
+    - uses: actions/setup-java@v3
+      with:
+        java-version: '17'
         distribution: 'temurin'
-    - uses: skjolber/maven-cache-github-action@v1
+    - uses: actions/cache@v3
+      with:
+        path: ~/.m2
+        key: ${{ runner.os }}-build-${{ hashFiles('**/pom.xml') }}
```

Comment:
So was the resolution that this `actions/cache` + the `maven-settings-action` need to be used together in order to get both caching and `ghcr.io`/GitHub Maven Packages working, rather than using the `cache:` configuration of `actions/setup-java`?

Reply:
yes- the reason was because we needed the `restore-keys` option that `actions/cache` gives us since cryostat and -core both have a pom.xml file so the `key` wouldn't produce a hit.

Reply:
Cache conflicts in a GitHub Actions CI pipeline. Fun.

Reply:
also, I tried to use the server setup options in `actions/setup-java`, but I don't think it was adding it to the right place? `maven-settings-action` adds the cryostat server to `settings.xml` so we kept it the way it is.